### PR TITLE
Fixed #19527 -- Allowed QuerySet.bulk_create() to set the primary key of its objects.

### PR DESCRIPTION
**Pull request by blablacio**
PostgreSQL supports RETURNING id when doing a bulk create. With this,
we can set the primary keys on the newly saved objects from the keys
that the database supplied.

Addresses #19527
Rebasing #5166 to Django 1.10

### CODE REVIEW

**Comment by timgraham** [comment]
There's another updated version here: https://github.com/django/django/pull/5917

**Comment by blablacio** [comment]
I noticed that, but was unsure if it's getting merged and you kind of accepted the change in #5166. Maybe you could shed some light on your plan for this feature as we depend on it?

**Comment by timgraham** [comment]
I don't have any particular plans other than to review pull requests as quickly as possible. If merging the other one first makes this implementation simpler, that sounds like a good thing.

**Comment by apollo13** [comment]
Could you check the SQL spec if "RETURNING id" for multiple rows is something other databases could support too? If yes, the code should get moved into the standard insert compiler and if no, you can at least drop part of the if statements, since you know that those flags will be true in pg always.

**Comment by timabbott** [comment]
There's discussion of how RETURNING ID works differently in other database engines (e.g. in Oracle) in the original ticket https://code.djangoproject.com/ticket/19527.

File: django/db/backends/postgresql/compiler.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+from django.db.models.sql import compiler
+
+SQLCompiler = compiler.SQLCompiler
+SQLDeleteCompiler = compiler.SQLDeleteCompiler
+SQLUpdateCompiler = compiler.SQLUpdateCompiler
+SQLAggregateCompiler = compiler.SQLAggregateCompiler
+
+
+class SQLInsertCompiler(compiler.SQLInsertCompiler):
```
Comment (akaariai) [inline django/db/backends/postgresql/compiler.py]: I don't like the duplication of code done here. Could we instead branch on database feature in the main compiler class?

Comment (blablacio) [inline django/db/backends/postgresql/compiler.py]: Sounds reasonable, I'll change it as requested.

**Comment by blablacio** [comment]
Requested change applied, some code cleaned up. Tests are passing fine now, there were some problems with previous commit.

**Comment by timabbott** [comment]
@blablacio looks like there are some new merge conflicts?

**Comment by blablacio** [comment]
@timgraham Just rebased onto master once again, hopefully we can get this merged for the next minor release.

File: docs/releases/1.9.txt [hunk 2]
```diff
@@ -1508,3 +1508,5 @@ removed in Django 1.9 (please see the :ref:`deprecation timeline
 * Private API ``django.test.utils.TestTemplateLoader`` is removed.
 
 * The ``django.contrib.contenttypes.generic`` module is removed.
+
+* QuerySet.bulk_create sets pk on objects when using the Postgresql backend.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: This won't be backported to 1.9.

File: django/db/models/query.py [hunk 3]
```diff
@@ -1051,10 +1058,20 @@ def _batched_insert(self, objs, fields, batch_size):
             return
         ops = connections[self.db].ops
         batch_size = (batch_size or max(ops.bulk_batch_size(fields, objs), 1))
+        ret = []
```
Comment (timgraham) [inline django/db/models/query.py]: use a more descriptive name

File: django/db/models/query.py [hunk 4]
```diff
@@ -1051,10 +1058,20 @@ def _batched_insert(self, objs, fields, batch_size):
...
         for batch in [objs[i:i + batch_size]
                       for i in range(0, len(objs), batch_size)]:
-            self.model._base_manager._insert(batch, fields=fields,
-                                             using=self.db)
+            if connections[self.db].features.can_return_ids_from_bulk_insert:
+                if len(objs) > 1:
+                    ret.extend(self.model._base_manager._insert(batch, fields=fields,
```
Comment (timgraham) [inline django/db/models/query.py]: use a single line or use hanging indent (we avoid non-4 space indents)

File: django/db/models/query.py [hunk 5]
```diff
@@ -1051,10 +1058,20 @@ def _batched_insert(self, objs, fields, batch_size):
...
+                                                                using=self.db, return_id=True))
+                if len(objs) == 1:
+                    ret.append(self.model._base_manager._insert(batch, fields=fields,
+                                                                using=self.db, return_id=True))
+            else:
+                self.model._base_manager._insert(batch, fields=fields,
```
Comment (timgraham) [inline django/db/models/query.py]: single line as above

File: docs/ref/models/querysets.txt [hunk 6]
```diff
@@ -1790,12 +1790,16 @@ This has a number of caveats though:
...
-  does not retrieve and set the primary key attribute, as ``save()`` does.
+  does not retrieve and set the primary key attribute, as ``save()`` does,
+  unless the db backend also has the feature ``can_return_ids_from_bulk_insert``
+  (currently ``postgresql``).
 * It does not work with many-to-many relationships.
 
 .. versionchanged:: 1.9
 
     Support for using ``bulk_create()`` with proxy models was added.
+    Support for setting primary keys on objects on bulk_create, when using
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: `.. versionchanged:: 1.10`

**Comment by timgraham** [comment]
Add a mention of the `can_return_ids_from_bulk_insert` feature to the "Database backends" section of the 1.10 release notes.

File: django/db/backends/postgresql/operations.py [hunk 7]
```diff
@@ -59,6 +59,14 @@ def datetime_trunc_sql(self, lookup_type, field_name, tzname):
     def deferrable_sql(self):
         return " DEFERRABLE INITIALLY DEFERRED"
 
+    def fetch_returned_insert_ids(self, cursor):
+        """
+        Given a cursor object that has just performed an INSERT...RETURNING
+        statement into a table that has an auto-incrementing ID, returns the
```
Comment (timgraham) [inline django/db/backends/postgresql/operations.py]: returns->return (use PEP257 verb style for new docstrings)

File: django/db/models/query.py [hunk 8]
```diff
@@ -411,17 +411,20 @@ def bulk_create(self, objs, batch_size=None):
         Inserts each of the instances into the database. This does *not* call
         save() on each of the instances, does not send any pre/post save
         signals, and does not set the primary key attribute if it is an
-        autoincrement field. Multi-table models are not supported.
+        autoincrement field (except on Postgres). Multi-table models are
```
Comment (timgraham) [inline django/db/models/query.py]: except if features.can_return_ids_from_bulk_insert=True

File: django/db/models/query.py (hunk 4 above)
Comment (timgraham) [inline django/db/models/query.py]: I wouldn't repeat `self.model._base_manager._insert(batch, fields=fields, ...` but just vary the extend/append action based on the number. Are both branches tested?

File: docs/ref/models/querysets.txt [hunk 9]
```diff
@@ -1790,12 +1790,16 @@ This has a number of caveats though:
   ``post_save`` signals will not be sent.
 * It does not work with child models in a multi-table inheritance scenario.
 * If the model's primary key is an :class:`~django.db.models.AutoField` it
-  does not retrieve and set the primary key attribute, as ``save()`` does.
+  does not retrieve and set the primary key attribute, as ``save()`` does,
+  unless the db backend also has the feature ``can_return_ids_from_bulk_insert``
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: unless the database backend supports it (currently PostgreSQL).

File: docs/ref/models/querysets.txt [hunk 10]
```diff
@@ -1790,12 +1790,16 @@ This has a number of caveats though:
...
+  (currently ``postgresql``).
 * It does not work with many-to-many relationships.
 
 .. versionchanged:: 1.9
 
     Support for using ``bulk_create()`` with proxy models was added.
+    Support for setting primary keys on objects on bulk_create, when using
+    Postgres, was added.
```
Comment (timgraham) [inline docs/ref/models/querysets.txt]: PostgreSQL

File: docs/ref/models/querysets.txt (hunk 6 above)
Comment (timgraham) [inline docs/ref/models/querysets.txt]:
```
``bulk_create()``
```

**Comment by timgraham** [comment]
Please check if any parts of https://github.com/django/django/pull/5917 are needed in your version of the patch.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by blablacio** [comment]
@timgraham Just rebased onto master and pushed the code addressing issues you outlined.

**Comment by timgraham** [comment]
merged in 04240b23658f8935bbfebacccc23b5e47a1d6c22, thanks!

**Comment by westurner** [comment]
Thanks!

**Pull request closed without merging**